### PR TITLE
Contributing lib/ansible/modules/network/cloudengine/ce_static_route.py module to manage HUAWEI data center CloudEngine

### PR DESCRIPTION
**Pull request by QijunPan**
add ce_static_route

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

ce_static_route.py
##### ANSIBLE VERSION

```
ansible 2.3.0
config file = /etc/ansible/ansible.cfg
configured module search path = ['/home/ansible-2.1.1.0/lib/ansible/modules/core/network/cloudengine', '/home/ansible-2.1.1.0/lib/ansible/module_utils', '/home/ansible-2.1.1.0/lib/ansible/modules', '/home/ansible-2.1.1.0', '/home/ansible-2.1.1.0/lib/ansible/utils/']
```

##### SUMMARY

```
TASK [Config an ipv4 static route 2.2.2.0/24 10GE1/0/1 3.3.3.1] ****************
task path: /usr1/code/OpenNESS/code/current/KVM/intg/test/testcases/test-ce_static_route.yml:15
Using module file /usr/local/lib/python2.7/site-packages/ansible-2.3.0-py2.7.egg/ansible/modules/core/network/cloudengine/ce_static_route.py
<ce_6850> ESTABLISH LOCAL CONNECTION FOR USER: root
<ce_6850> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1487053829.08-56081827527412 `" && echo ansible-tmp-1487053829.08-56081827527412="` echo $HOME/.ansible/tmp/ansible-tmp-1487053829.08-56081827527412 `" ) && sleep 0'
<ce_6850> PUT /tmp/tmpXYt433 TO /root/.ansible/tmp/ansible-tmp-1487053829.08-56081827527412/ce_static_route.py
<ce_6850> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1487053829.08-56081827527412/ /root/.ansible/tmp/ansible-tmp-1487053829.08-56081827527412/ce_static_route.py && sleep 0'
<ce_6850> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1487053829.08-56081827527412/ce_static_route.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1487053829.08-56081827527412/" > /dev/null 2>&1 && sleep 0'
changed: [ce_6850] => {
    "changed": true, 
    "end_state": {
        "sroute": [
            {
                "afType": "ipv4unicast", 
                "description": null, 
                "destVrfName": "_public_", 
                "ifName": "Invalid0", 
                "maskLength": "8", 
                "nexthop": "192.109.130.1", 
                "preference": "1", 
                "prefix": "192.0.0.0", 
                "tag": "None", 
                "topologyName": "base", 
                "vrfName": "_public_"
            }, 
            {
                "afType": "ipv4unicast", 
                "description": null, 
                "destVrfName": "_public_", 
                "ifName": "Invalid0", 
                "maskLength": "8", 
                "nexthop": "10.136.53.1", 
                "preference": "60", 
                "prefix": "10.0.0.0", 
                "tag": "None", 
                "topologyName": "base", 
                "vrfName": "_public_"
            }, 
            {
                "afType": "ipv4unicast", 
                "description": null, 
                "destVrfName": "_public_", 
                "ifName": "10GE1/0/1", 
                "maskLength": "24", 
                "nexthop": "3.3.3.1", 
                "preference": "60", 
                "prefix": "2.2.2.0", 
                "tag": "None", 
                "topologyName": "base", 
                "vrfName": "_public_"
            }
        ]
    }, 
    "existing": {
        "sroute": [
            {
                "afType": "ipv4unicast", 
                "description": null, 
                "destVrfName": "_public_", 
                "ifName": "Invalid0", 
                "maskLength": "8", 
                "nexthop": "192.109.130.1", 
                "preference": "1", 
                "prefix": "192.0.0.0", 
                "tag": "None", 
                "topologyName": "base", 
                "vrfName": "_public_"
            }, 
            {
                "afType": "ipv4unicast", 
                "description": null, 
                "destVrfName": "_public_", 
                "ifName": "Invalid0", 
                "maskLength": "8", 
                "nexthop": "10.136.53.1", 
                "preference": "60", 
                "prefix": "10.0.0.0", 
                "tag": "None", 
                "topologyName": "base", 
                "vrfName": "_public_"
            }
        ]
    }, 
    "invocation": {
        "module_args": {
            "aftype": "v4", 
            "auth_pass": null, 
            "authorize": false, 
            "description": null, 
            "destvrf": null, 
            "host": "ce_6850", 
            "mask": "24", 
            "next_hop": "3.3.3.1", 
            "nhp_interface": "10GE1/0/1", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 12345, 
            "pref": null, 
            "prefix": "2.2.2.2", 
            "provider": null, 
            "ssh_keyfile": null, 
            "state": "present", 
            "tag": null, 
            "timeout": 10, 
            "transport": null, 
            "use_ssl": false, 
            "username": "rootDC", 
            "validate_certs": true, 
            "vrf": null
        }, 
        "module_name": "ce_static_route"
    }, 
    "proposed": {
        "afType": "v4", 
        "destVrfName": "_public_", 
        "ifName": "10GE1/0/1", 
        "mask": "24", 
        "next_hop": "3.3.3.1", 
        "preference": 60, 
        "prefix": "2.2.2.0", 
        "state": "present", 
        "vrfName": "_public_"
    }, 
    "updates": [
        "ip route-static 2.2.2.0 255.255.255.0 10GE1/0/1 3.3.3.1"
    ]
}
```

### CODE REVIEW

File: lib/ansible/modules/network/cloudengine/ce_static_route.py [hunk 1]
```diff
@@ -0,0 +1,845 @@
...
+---
+module: ce_static_route
+version_added: "2.3"
+short_description: Manages static route configuration.
+description:
+    - Manages the static routes of Huawei CloudEngine switches.
+author: Yang yang (@CloudEngine-Ansible)
+notes:
+    - If no vrf is supplied, vrf is set to default.
+      If state=absent, the route will be removed, regardless of the
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_static_route.py]: `I(state=absent)`

File: lib/ansible/modules/network/cloudengine/ce_static_route.py [hunk 2]
```diff
@@ -0,0 +1,845 @@
...
+        prefix=dict(required=True, type='str'),
+        mask=dict(required=True, type='str'),
+        aftype=dict(choices=['v4', 'v6'], required=True),
+        next_hop=dict(required=False, type='str'),
+        nhp_interface=dict(required=False, type='str'),
+        vrf=dict(required=False, type='str'),
+        destvrf=dict(required=False, type='str'),
+        tag=dict(required=False, type='str'),
+        description=dict(required=False, type='str'),
+        pref=dict(required=False, type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_static_route.py:831]: For integers please:
use `type='int`
remove the `isdigit` check from `check_params`

File: lib/ansible/modules/network/cloudengine/ce_static_route.py [hunk 3]
```diff
@@ -0,0 +1,845 @@
...
+
+        self.module.exit_json(**self.results)
+
+
+def main():
+    """main"""
+
+    argument_spec = dict(
+        prefix=dict(required=True, type='str'),
+        mask=dict(required=True, type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_static_route.py:823]: For integers please:
use `type='int`
remove the `isdigit` check from `check_params`

File: lib/ansible/modules/network/cloudengine/ce_static_route.py [hunk 4]
```diff
@@ -0,0 +1,845 @@
...
+                        if static_ele.text is not None:
+                            static_info["tag"] = static_ele.text
+                        else:
+                            static_info["tag"] = "None"
+                self.static_routes_info["sroute"].append(static_info)
+
+    def check_params(self):
+        """check all input params"""
+        # prefix, mask, aftype, next_hop, state, check
+        if not self.prefix or not self.mask or not self.aftype or not self.state:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_static_route.py]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
mutually_exclusive
required_together
required_one_of
require_if
```

File: lib/ansible/modules/network/cloudengine/ce_static_route.py [hunk 5]
```diff
@@ -0,0 +1,845 @@
...
+            self.module.fail_json(
+                msg='Error: Prefix or mask or address family type or state must be set.')
+        if not self.next_hop and not self.nhp_interface:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_static_route.py]: There are some options to `AnsibleModule` which you may be able to use to validate options passed in.
Look at existing modules for examples:
```
mutually_exclusive
required_together
required_one_of
require_if
```

File: lib/ansible/modules/network/cloudengine/ce_static_route.py [hunk 6]
```diff
@@ -0,0 +1,845 @@
...
+            if self.next_hop:
+                if not is_valid_v6addr(self.next_hop):
+                    self.module.fail_json(
+                        msg='Error: The %s is not a valid address' % self.next_hop)
+
+        # description check
+        if self.description:
+            if not is_valid_description(self.description):
+                self.module.fail_json(
+                    msg='Error: Dsecription length should be 1 - 35,and can not contain "?".')
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_static_route.py]: missing space between `,` and `and`

File: lib/ansible/modules/network/cloudengine/ce_static_route.py [hunk 7]
```diff
@@ -0,0 +1,845 @@
...
+
+    argument_spec = dict(
+        prefix=dict(required=True, type='str'),
+        mask=dict(required=True, type='str'),
+        aftype=dict(choices=['v4', 'v6'], required=True),
+        next_hop=dict(required=False, type='str'),
+        nhp_interface=dict(required=False, type='str'),
+        vrf=dict(required=False, type='str'),
+        destvrf=dict(required=False, type='str'),
+        tag=dict(required=False, type='str'),
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_static_route.py:829]: For integers please:
use `type='int`
remove the `isdigit` check from `check_params`

File: lib/ansible/modules/network/cloudengine/ce_static_route.py (hunk 3 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_static_route.py:823]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_static_route.py (hunk 7 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_static_route.py:829]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_static_route.py (hunk 2 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_static_route.py:831]: Let's keep this type, avoid conversion

File: lib/ansible/modules/network/cloudengine/ce_static_route.py (hunk 5 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_static_route.py]: OK, thanks.

File: lib/ansible/modules/network/cloudengine/ce_static_route.py (hunk 6 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_static_route.py]: I'll fix it.

File: lib/ansible/modules/network/cloudengine/ce_static_route.py (hunk 4 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_static_route.py]: OK, thanks.

File: lib/ansible/modules/network/cloudengine/ce_static_route.py (hunk 1 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_static_route.py]: I'll fix it.  thank you.

**Review by gundalow — APPROVED** [review]

**Pull request merged**